saleor/graphql/discount/types/promotion_events.py

```diff
@@ -0,0 +1,193 @@
+    class Meta:
+        description = (
+            "History log of the promotion started event."
+            + ADDED_IN_315
+            + PREVIEW_FEATURE
+        )
+        interfaces = [relay.Node, PromotionEvent]
+        model = models.PromotionEvent
+        doc_category = DOC_CATEGORY_DISCOUNTS
+
+
+class PromotionEndedEvent(ModelObjectType[models.PromotionEvent]):
+    class Meta:
+        description = (
+            "History log of the promotion ended event." + ADDED_IN_315 + PREVIEW_FEATURE
+        )
+        interfaces = [relay.Node, PromotionEvent]
+        model = models.PromotionEvent
+        doc_category = DOC_CATEGORY_DISCOUNTS
+
+
+class PromotionRuleEvent(graphene.Interface):
+    class Meta:
+        description = (
+            "History log of the promotion event related to rule."
+            + ADDED_IN_315
+            + PREVIEW_FEATURE
+        )
+        interfaces = [relay.Node]
+        model = models.PromotionEvent
+        doc_category = DOC_CATEGORY_DISCOUNTS
+
+    rule_id = graphene.String(
+        description="The rule ID associated with the promotion event."
+    )
+
+    @staticmethod
+    def resolve_rule_id(root: models.PromotionEvent, _info):
+        return root.parameters.get("rule_id", None)
+
+
+class PromotionRuleCreatedEvent(ModelObjectType[models.PromotionEvent]):
+    class Meta:
+        description = (
+            "History log of the promotion rule created event."
+            + ADDED_IN_315
+            + PREVIEW_FEATURE
+        )
+        interfaces = [relay.Node, PromotionEvent, PromotionRuleEvent]
+        model = models.PromotionEvent
+        doc_category = DOC_CATEGORY_DISCOUNTS
+
+
+class PromotionRuleUpdatedEvent(ModelObjectType[models.PromotionEvent]):
+    class Meta:
+        description = (
+            "History log of the promotion rule created event."
+            + ADDED_IN_315
+            + PREVIEW_FEATURE
+        )
+        interfaces = [relay.Node, PromotionEvent, PromotionRuleEvent]
+        model = models.PromotionEvent
+        doc_category = DOC_CATEGORY_DISCOUNTS
+
+
+class PromotionRuleDeletedEvent(ModelObjectType[models.PromotionEvent]):
+    class Meta:
+        description = (
+            "History log of the promotion rule created event."
+            + ADDED_IN_315
+            + PREVIEW_FEATURE
+        )
+        interfaces = [relay.Node, PromotionEvent, PromotionRuleEvent]
+        model = models.PromotionEvent
+        doc_category = DOC_CATEGORY_DISCOUNTS
+
+
+PROMOTION_EVENT_MAP = {
+    events.PROMOTION_CREATED: PromotionCreatedEvent,
+    events.PROMOTION_UPDATED: PromotionUpdatedEvent,
+    events.PROMOTION_STARTED: PromotionStartedEvent,
+    events.PROMOTION_ENDED: PromotionEndedEvent,
+    events.RULE_CREATED: PromotionRuleCreatedEvent,
+    events.RULE_UPDATED: PromotionRuleUpdatedEvent,
+    events.RULE_DELETED: PromotionRuleDeletedEvent,
+}
+
+
+class UnionPromotionEvent(graphene.Union):
```

Comment:
Where do you use this? 
Also, wouldn't be better to have `PromotionEventInterface` and rename `UnionPromotionEvent` to `PromotionEvent`. Then connect `PromotionEvent` (union) to the `Promotion`? 
Then you will have connected `union` as `Promotion.event`. It should be much more easier to use. 
(You will be able to make something like this:
```graphql
events {
    ...on PromotionUpdatedEvent{
        id
    }
    ...on PromotionStartedEvent{
        id
   }
```

Currently, as far as I understand the tests, you can't do that ☝️ , you have generic separation based on the interface. 
```graphql
    query Promotion($id: ID!) {
         promotion(id: $id) {
             id
             events {
                 ... on PromotionEvent {
                     type
                     createdBy {
                         ... on User {
                             id
                         }
                     }
                 }
                 ... on PromotionRuleEvent {
                     ruleId
                 }
             }
         }
     }
```
And this is the reason why you had to include `PROMOTION_EVENT_MAP` in `api.py`. You don't use it currently.